lang/rust/avro_derive/src/lib.rs

```diff
@@ -196,6 +283,252 @@ fn get_data_struct_schema_def(
+        },
+        ValueExpr::Option(value_variant) => {
+            let inner_expr = data_struct_value_field_as_value(
+                proc_macro2::Ident::new("item", span),
+                None,
+                *value_variant,
+                span,
+            );
+            match receiver_ident {
+                Some(recv) => quote!(match #recv.#ident {
+                    Some(item) => apache_avro::types::Value::Union(1, Box::new(#inner_expr)),
+                    None => apache_avro::types::Value::Union(0, Box::new(apache_avro::types::Value::Null)),
+                }),
+                None => quote!(match #ident {
+                    Some(item) => apache_avro::types::Value::Union(1, Box::new(#inner_expr)),
+                    None => apache_avro::types::Value::Union(0, Box::new(apache_avro::types::Value::Null)),
+                }),
+            }
+        }
+        ValueExpr::Array(value_variant) => {
+            let inner_expr = data_struct_value_field_as_value(
+                proc_macro2::Ident::new("item", span),
+                None,
+                *value_variant,
+                span,
+            );
+            match receiver_ident {
+                Some(recv) => {
+                    quote!(apache_avro::types::Value::Array(#recv.#ident.map(|item| #inner_expr)))
+                }
+                None => quote!(apache_avro::types::Value::Array(#ident.map(|item| #inner_expr))),
+            }
+        }
+        ValueExpr::HashMap(value_variant) => {
+            let inner_expr = data_struct_value_field_as_value(
+                proc_macro2::Ident::new("value", span),
+                None,
+                *value_variant,
+                span,
+            );
+            match receiver_ident {
+                Some(recv) => quote!(apache_avro::types::Value::Map(
+                    #recv.#ident
+                        .map(|(key, value)| (key, #inner_expr))
+                        .collect::<HashMap<String, apache_avro::types::Value>>()
+                )),
+                None => quote!(apache_avro::types::Value::Map(
+                    #ident
+                        .map(|(key, value)| (key, #inner_expr))
+                        .collect::<HashMap<String, apache_avro::types::Value>>()
+                )),
+            }
+        }
+    }
+}
+
+fn data_struct_value_assignment(
+    field: &Field,
+    index: usize,
+) -> Result<TokenStream, Vec<syn::Error>> {
+    let field_ident = field.clone().ident.unwrap();
+    let ident = proc_macro2::Ident::new("value", field.span());
+    let value_expr = type_to_value_expr(&field.ty)?;
+    let value_assignment =
+        data_struct_value_field_as_assigned(quote!(#ident), None, value_expr, ident.span());
+    Ok(quote!(let #field_ident = match record_fields.get(#index) {
+        Some((_, value)) => {
+            #value_assignment
+        },
+        _ => return Err(Self::Error::GetField(format!("index: {}", #index)))
```

Comment:
```suggestion
        _ => return Err(Self::Error::GetField(format!("index: {#index}", )))
```

the linter would complain about this code if it was hand-written, not use if it complains about macro code...